ufl/algorithms/apply_derivatives.py

```diff
@@ -657,16 +670,51 @@ def reference_value(self, o):
         """Differentiate a reference_value."""
         # grad(o) == grad(rv(f)) -> K_ji*rgrad(rv(f))_rj
         f = o.ufl_operands[0]
-        if isinstance(f.ufl_element().pullback, PhysicalPullback):
-            # TODO: Do we need to be more careful for immersed things?
-            return ReferenceGrad(o)
-
         if not f._ufl_is_terminal_:
             raise ValueError("ReferenceValue can only wrap a terminal")
-        domain = extract_unique_domain(f)
-        K = JacobianInverse(domain)
-        Do = grad_to_reference_grad(o, K)
-        return Do
+        domain = extract_unique_domain(f, expand_mixed_mesh=False)
+        if isinstance(domain, MixedMesh):
+            element = f.ufl_function_space().ufl_element()
+            if element.num_sub_elements != len(domain):
+                raise RuntimeError(f"{element.num_sub_elements} != {len(domain)}")
+            g = ReferenceGrad(o)
+            ref_dim = g.ufl_shape[-1]
+            components = []
+            dofoffset = 0
+            for d, e in flatten_domain_element(domain, element):
+                esh = e.reference_value_shape
+                if esh == ():
+                    esh = (1,)
+                if isinstance(e.pullback, PhysicalPullback):
+                    if ref_dim != self._var_shape[0]:
+                        raise NotImplementedError("""
+                            PhysicalPullback not handled for immersed domain :
+                            reference dim ({ref_dim}) != physical dim (self._var_shape[0])""")
+                    for idx in range(int(np.prod(esh))):
+                        for i in range(ref_dim):
+                            components.append(g[(dofoffset + idx,) + (i,)])
+                else:
+                    K = JacobianInverse(d)
+                    rdim, gdim = K.ufl_shape
+                    if rdim != ref_dim:
+                        raise RuntimeError(f"{rdim} != {ref_dim}")
+                    if gdim != self._var_shape[0]:
+                        raise RuntimeError(f"{gdim} != {self._var_shape[0]}")
+                    for idx in range(int(np.prod(esh))):
+                        for i in range(gdim):
+                            temp = Zero()
+                            for j in range(rdim):
+                                temp += g[(dofoffset + idx,) + (j,)] * K[j, i]
+                            components.append(temp)
+                dofoffset += int(np.prod(esh))
```

Comment:
Some documentation of the logic within this loop would be nice.
I guess the idea is:
- For a mixed domain problem, accumulate all derivatives in expanded index notation.
Please correct me if I am mistaken.

I think one thing that would help is to define `np.prod(esh)` as a variable, as it is used in multiple loops and to accumulate the counter.

Also, I think you do not need the
```
if esh == ():
    esh = (1,)
```
as `int(np.prod(()))` ref: https://numpy.org/doc/2.1/reference/generated/numpy.prod.html
> The product of an empty array is the neutral element 1

One could add a guard against future changing behavior by adding
```python
esh = e.reference_value_shape
num_components = int(np.prod(esh))
assert num_components > 0
```

Reply:
You are right. Simplified as suggested and added more comments. Thanks.